src/config.py

```diff
@@ -0,0 +1,222 @@
+
+        Returns:
+            List of properties to be set
+        """
+        broker_id = self.charm.unit.name.split("/")[1]
+        host = (
+            self.charm.model.get_relation(PEER).data[self.charm.unit].get("private-address", None)
+        )
+        return [
+            f"broker.id={broker_id}",
+            f"advertised.listeners=SASL_PLAINTEXT://{host}:9092",
+            f'zookeeper.connect={self.zookeeper_config["connect"]}',
+            f'listener.name.sasl_plaintext.scram-sha-512.sasl.jaas.config=org.apache.kafka.common.security.scram.ScramLoginModule required username="sync" password="{self.sync_password}";',
+        ]
+
+    @property
+    def super_users(self) -> str:
+        """Generates all users with super/admin permissions for the cluster from relations.
+
+        Formatting allows passing to the `super.users` property.
+
+        Returns:
+            Semi-colon delimited string of current super users
+        """
+        super_users = ["sync"]
+        for relation in self.charm.model.relations[REL_NAME]:
+            extra_user_roles = relation.data[relation.app].get("extra-user-roles", "")
+            password = (
+                self.charm.model.get_relation(PEER)
+                .data[self.charm.app]
+                .get(f"relation-{relation.id}", None)
+            )
+            # if passwords are set for client admins, they're good to load
+            if "admin" in extra_user_roles and password is not None:
+                super_users.append(f"relation-{relation.id}")
+
+        super_users_arg = [f"User:{user}" for user in super_users]
+
+        return ";".join(super_users_arg)
+
+    @property
+    def server_properties(self) -> List[str]:
+        """Builds all properties necessary for starting Kafka service.
+
+        This includes charm config, replication, SASL/SCRAM auth and default properties.
+
+        Returns:
+            List of properties to be set
+        """
+        return (
+            [
+                f"offsets.retention.minutes={self.charm.config['offsets-retention-minutes']}",
+                f"log.retention.hours={self.charm.config['log-retention-hours']}",
+                f"auto.create.topics={self.charm.config['auto-create-topics']}",
+                f"super.users={self.super_users}",
+            ]
+            + self.default_replication_properties
+            + self.auth_properties
+            + DEFAULT_CONFIG_OPTIONS.split("\n")
+        )
+
+    def set_jaas_config(self) -> None:
+        """Writes the Kafka JAAS config using ZooKeeper relation data."""
+        jaas_config = f"""
+            Client {{
+                org.apache.zookeeper.server.auth.DigestLoginModule required
+                username="{self.zookeeper_config['username']}"
+                password="{self.zookeeper_config['password']}";
+            }};
+        """
+        safe_write_to_file(content=jaas_config, path=self.jaas_filepath, mode="w")
```

Comment:
can this error?

Reply:
It can I suppose, as it's calling `open()` and `write()`, but if it did error, I'd want that to block the charm.